### PR TITLE
4.0.8 version upgrade tip

### PR DESCRIPTION
**Pull request by margud**

### CODE REVIEW

File: src/Service/UpgradeNoticeService.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace Mollie\Service;
+
+use Mollie\Config\Config;
+
+class UpgradeNoticeService
+{
+    public function isUpgradeNoticeClosed($timeStamp)
```
Comment (tomas862) [inline src/Service/UpgradeNoticeService.php:9]: comment typehints

File: controllers/admin/AdminMollieAjaxController.php [hunk 2]
```diff
@@ -94,4 +98,9 @@ private function testApiKeys()
             ]
         ));
     }
+
+    private function closeUpgradeNotice()
+    {
+        Configuration::updateValue(Config::MODULE_UPGRADE_NOTICE_CLOSE_DATE, time());
```
Comment (tomas862) [inline controllers/admin/AdminMollieAjaxController.php:104]: as I always mentioned - think always about interfaces and code testability . ClockInterface with method getNowTs() meaning gets current time as timestamp

File: src/Service/UpgradeNoticeService.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        $closedTimeStamp = \Configuration::get(Config::MODULE_UPGRADE_NOTICE_CLOSE_DATE);
+        $closeDuration = strtotime(Config::MODULE_UPGRADE_NOTICE_CLOSE_DURATION) - time();
+        if ($closedTimeStamp + $closeDuration > $timeStamp) {
+            return true;
+        }
+
+        return false;
+    }
```
Comment (tomas862) [inline src/Service/UpgradeNoticeService.php:18]: do you trust your code? Well I dont :D  add unit test for this please

File: src/Service/UpgradeNoticeService.php [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace Mollie\Service;
+
+use Mollie\Config\Config;
+
+class UpgradeNoticeService
+{
```
Comment (tomas862) [inline src/Service/UpgradeNoticeService.php:8]: dockblocks for typehinting

File: src/Service/SettingsSaveService.php [hunk 5]
```diff
@@ -182,6 +182,23 @@ public function saveSettings(&$errors = [])
...
         }
 
+        $apiKey = (int)$environment === Config::ENVIRONMENT_LIVE ?
+            $mollieApiKey : $mollieApiKeyTest;
+
+        if ($apiKey) {
+            try {
+                $api = $this->apiService->setApiKey($apiKey, $this->module->version);
+                if ($api === null) {
+                    throw new Exception('Failed to connect to mollie API');
```
Comment (tomas862) [inline src/Service/SettingsSaveService.php:192]: well dont throw generic Exceptions, create your own. MollieApiException or smth

File: src/Config/Config.php [hunk 6]
```diff
@@ -231,6 +231,9 @@ class Config
 
     const CARTES_BANCAIRES = 'cartesbancaires';
 
+    const MODULE_UPGRADE_NOTICE_CLOSE_DATE = 'MODULE_UPGRADE_NOTICE_CLOSE_DATE';
+    const MODULE_UPGRADE_NOTICE_CLOSE_DURATION = '28 days';
```
Comment (JevgenijVisockij) [inline src/Config/Config.php:235]: Const name is okay, but configs name should be something unique to your module. For example MOLLIE_MODULE_UPGRADE_NOTICE_CLOSE_DATE 

There is always a chance that some other module decides to name their config 'MODULE_UPGRADE_NOTICE_CLOSE_DATE'

**Pull request merged**